### PR TITLE
domain/affinity: update pd client for api

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/64938
Problem Summary:

### Check List

Tests 

- [x] Unit test
- [x] Integration test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

## Summary by CodeRabbit

* **Performance**
  * Faster, more scalable retrieval of affinity groups for large ID requests.
* **Bug Fixes**
  * Reduced failures by avoiding oversized queries and adding robust fallbacks when the backend rejects ID-based fetches.
  * More reliable affinity group creation with idempotent checks and safe fallback creation paths.
* **New Features**
  * Support for forceful batch deletion of affinity groups.
* **Tests**
  * Added comprehensive tests for creation, retrieval, fallbacks, and boundary cases.
* **Chores**
  * Bumped an internal backend dependency to a newer version.

### CODE REVIEW

File: pkg/domain/affinity/manager.go [hunk 1]
```diff
@@ -90,20 +68,48 @@ func (m *pdManager) GetAffinityGroups(ctx context.Context, ids []string) (map[st
...
+	for i, id := range ids {
+		if i == 0 {
+			total += len("ids=")
+		} else {
+			total += len("&ids=")
+		}
+		total += len(id)
+	}
+	return total
+}
```
[the remark this answers is not in the document]
Comment (lhy1024) [reply pkg/domain/affinity/manager.go]: No code change needed here. Affinity group IDs are internally generated in TiDB (`_tidb_t_%d` / `_tidb_pt_%d_p%d`) and only contain URL-safe characters, so raw length equals escaped length in our current call paths.

**Review by cfzjywxk — APPROVED** [review]

File: pkg/domain/affinity/manager.go [hunk 2]
```diff
@@ -45,34 +50,7 @@ func (m *pdManager) CreateAffinityGroupsIfNotExists(ctx context.Context, groups
...
-	}
-
-	// If all groups already exist, return success
-	if len(groupsToCreate) == 0 {
-		return nil
-	}
-
-	// Create only the groups that don't exist
-	_, err = m.Client.CreateAffinityGroups(ctx, groupsToCreate)
+	_, err := m.Client.CreateAffinityGroups(ctx, groups, pdhttp.WithSkipExistCheck())
```
Comment (D3Hunter) [inline pkg/domain/affinity/manager.go:64]: #### ⚠️ **[Major]** Idempotent affinity-group creation now depends on new PD query support without a fallback

**Why**
`CreateGroupsIfNotExists` is a critical DDL retry path, but the new implementation removed client-side existence filtering and now unconditionally relies on `skip_exist_check` support in PD.

**Scope**
pkg/domain/affinity/manager.go:53

**Risk if unchanged**
In N-1 mixed-version or partially upgraded PD deployments, retries after partial success can fail with duplicate-group errors, causing CREATE/ALTER/TRUNCATE affinity-related DDL jobs to fail instead of being idempotent.

**Evidence**
This diff replaces the previous `GetAffinityGroups` + filter-before-create flow with a single call: `m.Client.CreateAffinityGroups(ctx, groups, pdhttp.WithSkipExistCheck())`, and no compatibility fallback remains in `CreateAffinityGroupsIfNotExists`.

**Change request**
is this backward compatible with N-1 nodes? Please add a fallback path: if `skip_exist_check` is unsupported or rejected, fall back to the prior behavior (query existing IDs and only create missing groups) to preserve retry idempotency.

Comment (D3Hunter) [inline pkg/domain/affinity/manager.go:64]: #### ⚠️ **[Major]** Create-idempotency now depends on PD skip-exist feature without compatibility guard

**Why**
The retry-safe behavior for DDL affinity-group creation was previously enforced client-side by pre-checking existing groups. The new flow unconditionally depends on server-side `skip_exist_check` support, but there is no capability/version gate or fallback branch if that semantic is unavailable.

**Scope**
pkg/domain/affinity/manager.go:53

**Risk if unchanged**
On clusters where PD does not honor `skip_exist_check`, retried DDL operations can fail with duplicate-group errors, turning a recoverable retry path into a user-visible DDL failure.

**Evidence**
`CreateAffinityGroupsIfNotExists` now always calls `CreateAffinityGroups(..., pdhttp.WithSkipExistCheck())` and removed the former read-then-filter idempotency path from this method. No compatibility check or fallback path is added in `pkg/domain/affinity`.

**Change request**
is this backward compatible with N-1 nodes? Please add a capability/version gate (or fallback to the prior pre-check path on unsupported PD responses) so retry idempotency does not regress across mixed-version deployments.

Comment (D3Hunter) [inline pkg/domain/affinity/manager.go:64]: #### ⚠️ **[Major]** Affinity create/get paths assume new PD query semantics without N-1 fallback

**Why**
The new manager path unconditionally relies on query-based affinity APIs introduced by the upgraded `github.com/tikv/pd/client`, but there is no version gate or fallback for mixed-version clusters. On older PD handlers, create still rejects existing group IDs and GET only exposes the all-groups endpoint semantics.

**Scope**
pkg/domain/affinity/manager.go:53

**Risk if unchanged**
During rolling upgrade or rollback windows with older PD nodes, idempotent DDL retries can fail with duplicate-group errors, and targeted affinity-group reads can behave differently than expected. That can break CREATE/ALTER/TRUNCATE retry safety and operator expectations during upgrade sequencing.

**Evidence**
`pkg/domain/affinity/manager.go:53` now always calls `CreateAffinityGroups(..., pdhttp.WithSkipExistCheck())`; `pkg/domain/affinity/manager.go:71` routes most reads to `Client.GetAffinityGroups(ctx, ids)`. In the upgraded PD client module (`github.com/tikv/pd/client@v0.0.0-20260228084044-4f5039d43753/http/interface.go:1291` and `:1337`), those calls become `POST /pd/api/v2/affinity-groups?skip_exist_check=true` and `GET /pd/api/v2/affinity-groups?ids=...`. In an N-1 PD server snapshot (`github.com/tikv/pd@v1.1.0-beta.0.20260123090650-b6bc795a6005/server/apiv2/handlers/affinity.go:111` and `:415`), POST query handling recognizes only `?delete`, create explicitly rejects existing groups (`:147-149`), and GET handler is all-groups oriented without IDs filtering.

**Change request**
`is this backward compatible with N-1 nodes?` Add a capability/version guard or a runtime fallback: if PD does not support `skip_exist_check` or `ids` filtering, fall back to the legacy pre-check/filter flow (GetAll+local filter before create) so retry behavior stays upgrade-safe.

File: pkg/domain/affinity/manager.go [hunk 3]
```diff
@@ -90,20 +68,48 @@ func (m *pdManager) GetAffinityGroups(ctx context.Context, ids []string) (map[st
...
+			return nil, err
+		}
+		return filterAffinityGroups(allGroups, ids), nil
 	}
+	return m.Client.GetAffinityGroups(ctx, ids)
+}
 
-	// Filter by requested IDs
-	result := make(map[string]*pdhttp.AffinityGroupState)
+func affinityGroupIDsQueryLen(ids []string) int {
```
Comment (D3Hunter) [inline pkg/domain/affinity/manager.go]: #### 🟡 **[Minor]** `affinityGroupIDsQueryLen` implies exact query size, but current logic is only an estimate

**Why**
The helper name reads as an exact URL-query-length calculation, but the implementation sums raw `len(id)` values and does not account for URL escaping. This weakens behavior-signaling clarity for a limit-enforcement branch.

**Scope**
pkg/domain/affinity/manager.go:81

**Risk if unchanged**
Future maintainers may treat the threshold gate as precise and rely on it for safety, which can make boundary behavior around query-size limits harder to reason about and maintain.

**Evidence**
`affinityGroupIDsQueryLen` adds `len("ids=")`, `len("&ids=")`, and `len(id)`, while PD client query construction escapes IDs via `url.QueryEscape(id)` before building the request URI.

**Change request**
Prefer `estimatedAffinityGroupIDsQueryLen`; current name is confusing. Alternatively, compute length using escaped IDs so the function name and behavior stay aligned.

File: pkg/domain/affinity/manager.go [hunk 4]
```diff
@@ -32,6 +32,11 @@ type pdManager struct {
 	pdhttp.Client
 }
 
+const (
```
Comment (D3Hunter) [inline pkg/domain/affinity/manager.go:42]: #### 🟡 **[Minor]** Fallback branch uses hard thresholds without intent documentation

**Why**
The new `GetAllAffinityGroups` fallback is non-obvious behavior (switching from targeted fetch to full-list fetch + filter), but there is no rationale comment explaining why `100` IDs and `4096` query length were chosen.

**Scope**
pkg/domain/affinity/manager.go:35

**Risk if unchanged**
Thresholds can be changed or copied without understanding protocol and performance tradeoffs, increasing maintenance risk and making future tuning error-prone.

**Evidence**
Constants `maxAffinityGroupIDsQueryLen` and `maxAffinityGroupIDsCount` are introduced and immediately used in a branch that changes retrieval strategy, with no nearby why-comment or source-limit reference.

**Change request**
Please add a short why-comment for this branch, including where `4096` and `100` come from and why full-scan fallback is the intended behavior.

File: pkg/domain/affinity/manager.go [hunk 5]
```diff
@@ -90,20 +68,48 @@ func (m *pdManager) GetAffinityGroups(ctx context.Context, ids []string) (map[st
 		return make(map[string]*pdhttp.AffinityGroupState), nil
 	}
 
-	// TODO: avoid using GetAllAffinityGroups
-	allGroups, err := m.Client.GetAllAffinityGroups(ctx)
-	if err != nil {
-		return nil, err
+	if len(ids) > maxAffinityGroupIDsCount || affinityGroupIDsQueryLen(ids) > maxAffinityGroupIDsQueryLen {
```
Comment (D3Hunter) [inline pkg/domain/affinity/manager.go]: #### 🟡 **[Minor]** URI-length guard can miss over-limit requests because it does not model escaped query size

**Why**
The fallback decision uses raw string lengths, but the PD HTTP client URL-escapes each ID before building the query. This can under-estimate transmitted query size and keep the code on the direct `GetAffinityGroups` path when it should fall back.

**Scope**
pkg/domain/affinity/manager.go:71

**Risk if unchanged**
Boundary ID sets can still hit request-line/query limits and fail intermittently with URI-too-long style errors, weakening lookup reliability for large or encoded ID inputs.

**Evidence**
`affinityGroupIDsQueryLen` sums `len(id)` directly, while the upgraded PD client encodes each ID via `url.QueryEscape` in `http/interface.go` before issuing `GET /pd/api/v2/affinity-groups?...`.

**Change request**
add a case for encoded/edge IDs near the query-length boundary, and compute the guard using the same escaped form as the actual request builder so fallback behavior is deterministic.

Comment (D3Hunter) [inline pkg/domain/affinity/manager.go]: #### 🟡 **[Minor]** No targeted tests for new compatibility branches and fallback thresholds

**Why**
The patch changes core affinity behavior (server-side skip-existing create and conditional ID-filter query path) but does not add focused tests for mixed-version behavior, query-length/count branch selection, or fallback guarantees.

**Scope**
pkg/domain/affinity/manager.go:71

**Risk if unchanged**
Compatibility regressions can slip through CI and only surface during upgrade/rollback operations, where retry failures are costly and hard to diagnose.

**Evidence**
`65592.diff` modifies dependency versions plus `pkg/domain/affinity/manager.go` logic, including new branches (`affinityGroupIDsQueryLen`, `filterAffinityGroups`, and direct `GetAffinityGroups` path), but no `_test.go` changes are included for these behaviors.

**Change request**
`add UT for it` and `please add an integration test for this path`: cover (1) unsupported/new-PD capability matrix for create/get, (2) ids-count/query-length boundary behavior around 100 and 4096 thresholds, and (3) retry idempotency when groups already exist.

**Review by D3Hunter — COMMENTED** [review]
**AI-generated review based on @D3Hunter's standards; manual review will be performed after all comments are resolved.**

### Summary

- Total findings: 8
- Inline comments: 8
- Summary-only findings (no inline anchor): 0

### Findings (highest risk first)

#### ⚠️ **[Major]** (3)

1. Idempotent affinity-group creation now depends on new PD query support without a fallback (`pkg/domain/affinity/manager.go:53`)
2. Create-idempotency now depends on PD skip-exist feature without compatibility guard (`pkg/domain/affinity/manager.go:53`)
3. Affinity create/get paths assume new PD query semantics without N-1 fallback (`pkg/domain/affinity/manager.go:53`)

#### 🟡 **[Minor]** (4)

1. `affinityGroupIDsQueryLen` implies exact query size, but current logic is only an estimate (`pkg/domain/affinity/manager.go:81`)
2. Fallback branch uses hard thresholds without intent documentation (`pkg/domain/affinity/manager.go:35`)
3. URI-length guard can miss over-limit requests because it does not model escaped query size (`pkg/domain/affinity/manager.go:71`)
4. No targeted tests for new compatibility branches and fallback thresholds (`pkg/domain/affinity/manager.go:71`)

#### ℹ️ **[Info]** (1)

1. `CreateAffinityGroupsIfNotExists` comment does not describe the current single-call path (`pkg/domain/affinity/manager.go:45`)
